File: learn-wire-api-service/src/main/java/com/learn/wire/dto/study/query/StudyMode.java
package com.learn.wire.dto.study.query;

import com.learn.wire.constant.StudyConst;
import com.learn.wire.exception.BadRequestException;

public enum StudyMode {
    REVIEW(StudyConst.MODE_REVIEW),
    MATCH(StudyConst.MODE_MATCH),
    GUESS(StudyConst.MODE_GUESS),
    RECALL(StudyConst.MODE_RECALL),
    FILL(StudyConst.MODE_FILL);

    private final String value;

    StudyMode(String value) {
        this.value = value;
    }

    public String value() {
        return this.value;
    }

    public static StudyMode fromValue(String rawValue) {
        final String normalized = rawValue == null
                ? ""
                : rawValue.trim();
        for (final StudyMode candidate : values()) {
            if (candidate.value.equalsIgnoreCase(normalized)) {
                return candidate;
            }
        }
        throw new BadRequestException(StudyConst.MODE_INVALID_KEY);
    }
}

File: lib/features/flashcards/study/model/study_mode.dart
import '../../../../core/utils/string_utils.dart';
import 'study_constants.dart';

enum StudyMode { review, match, guess, recall, fill }

extension StudyModeApiX on StudyMode {
  String get apiValue {
    return switch (this) {
      StudyMode.review => StudyConstants.modeReview,
      StudyMode.match => StudyConstants.modeMatch,
      StudyMode.guess => StudyConstants.modeGuess,
      StudyMode.recall => StudyConstants.modeRecall,
      StudyMode.fill => StudyConstants.modeFill,
    };
  }

  static StudyMode fromApiValue(String rawValue) {
    final String normalized = StringUtils.normalize(rawValue).toLowerCase();
    if (normalized == StudyConstants.modeReview) {
      return StudyMode.review;
    }
    if (normalized == StudyConstants.modeMatch) {
      return StudyMode.match;
    }
    if (normalized == StudyConstants.modeGuess) {
      return StudyMode.guess;
    }
    if (normalized == StudyConstants.modeRecall) {
      return StudyMode.recall;
    }
    if (normalized == StudyConstants.modeFill) {
      return StudyMode.fill;
    }
    throw UnsupportedError(
      '${StudyConstants.unsupportedModeMessagePrefix}$rawValue',
    );
  }
}
